src/annotator/highlight-clusters.tsx

```diff
@@ -121,6 +141,105 @@ export class HighlightClusterController implements Destroyable {
+      return;
+    }
+
+    this._updateHighlightData(getHighlights(this._element));
+    this._updateSVGHighlightOrdering();
+  }
+
+  /**
+   * Walk a tree of <hypothesis-highlight> elements, adding `data-nesting-level`
+   * and `data-cluster-level` data attributes to <hypothesis-highlight>s and
+   * their associated SVG highlight (<rect>) elements.
+   *
+   * - `data-nesting-level` - generational depth of the applicable
+   *   `<hypothesis-highlight>` relative to outermost `<hypothesis-highlight>`.
+   * - `data-cluster-level` - number of `<hypothesis-highlight>` generations
+   *   since the cluster value changed.
+   *
+   * @param highlightEls - A collection of sibling <hypothesis-highlight>
+   * elements
+   * @param parentCluster - The cluster value of the parent highlight to
+   * `highlightEls`, if any
+   * @param nestingLevel - The nesting "level", relative to the outermost
+   * <hypothesis-highlight> element (0-based)
+   * @param parentClusterLevel - The parent's nesting depth, per its cluster
+   * value (`parentCluster`). i.e. How many levels since the cluster value
+   * changed? This allows for nested styling of highlights of the same cluster
+   * value.
+   */
+  _updateHighlightData(
+    highlightEls: HighlightElement[],
+    parentCluster = '',
+    nestingLevel = 0,
+    parentClusterLevel = 0
+  ) {
+    Array.from(highlightEls).forEach(hEl => {
+      const elCluster =
+        clusterValues.find(cv => hEl.classList.contains(cv)) ?? 'other-content';
+
+      const elClusterLevel =
+        parentCluster && elCluster === parentCluster
+          ? parentClusterLevel + 1
+          : 0;
+
+      hEl.setAttribute('data-nesting-level', `${nestingLevel}`);
+      hEl.setAttribute('data-cluster-level', `${elClusterLevel}`);
+
+      if (hEl.svgHighlight) {
+        hEl.svgHighlight.setAttribute('data-nesting-level', `${nestingLevel}`);
+        hEl.svgHighlight.setAttribute(
+          'data-cluster-level',
+          `${elClusterLevel}`
+        );
+      }
+
+      this._updateHighlightData(
+        getHighlights(hEl),
+        elCluster /* parentCluster */,
+        nestingLevel + 1 /* nestingLevel */,
+        elClusterLevel /* parentClusterLevel */
+      );
+    });
+  }
+
+  /**
+   * Ensure that SVG <rect> elements are ordered correctly: inner (nested)
+   * highlights should be visible on top of outer highlights.
+   *
+   * All SVG <rect>s drawn for a PDF page are siblings. To ensure that the
+   * <rect>s associated with outer highlights don't show up on top of (and thus
+   * obscure) nested highlights, order the <rects> by their `data-nesting-level`
+   * value if they are not already.
+   */
+  _updateSVGHighlightOrdering() {
+    const nestingLevel = (el: Element) =>
+      parseInt(el.getAttribute('data-nesting-level') ?? '0', 10);
+
+    for (const [layer, layerHighlights] of getSVGHighlights()) {
+      const correctlyOrdered = layerHighlights.every((svgEl, idx, allEls) => {
```

Comment:
This loop could be flipped around to bail when the first pair of elements that are not in order are encountered using `.some`, whereas `.every` will always test every element. An alternate and incompatible optimization would be to compute the nesting level once per element `layerHighlights.map(nestingLevel)` rather than twice.

I don't know how many elements you'd need before this would make a measurable difference, probably quite a lot.